lib/rules/no-unused-vars.js

```diff
@@ -126,6 +131,30 @@ module.exports = {
         }
 
         /**
+         * Determines if a variable has a sibling rest property
+         * @param {Variable} variable - EScope variable object.
+         * @returns {boolean} True if the variable is exported, false if not.
+         * @private
+         */
+        function hasRestSpreadSibling(variable) {
+            if (config.ignoreRestSiblings) {
+                const restProperties = new Set(["ExperimentalRestProperty", "RestProperty"]);
+
+                return variable.defs
+                    .filter(def => def.name.type === "Identifier")
+                    .some(def => (
+                        def.node.id &&
+                        def.node.id.type === "ObjectPattern" &&
+                        def.node.id.properties.length &&
+                        restProperties.has(def.node.id.properties[def.node.id.properties.length - 1].type) &&  // last property is a rest property
+                        def.node.id.properties.some(prop => prop.type === "Property" && prop.key.name === variable.name)  // variable is sibling of the rest property
```

Comment:
I don't think comparing `prop.key.name` is correct, because the property names and the variable names are independent from each other:

```js
// This will get reported, because prop.key.name is "foo" and variable.name is "_"
var { foo: _, ...bar } = baz;
doSomethingWith(bar);
```

I think the easiest way to do this would be to replace the "variable is sibling of the rest property" line with something using [`def.name`](https://estools.github.io/escope/Variable.html#defs) property, which is a reference to the `Identifier` node for the given variable. That way, we also wouldn't have to iterate over all the properties:

```js
// verify that this definition is not part of a RestProperty
!restProperties.has(def.name.parent.type)
```